### DISCUSSION TITLE
OPENTELEMETRY Metric Data support

### DISCUSSION THREAD

**Discussion by suzyw-w** [Kong/kong#13620] [Ideas and feature requests]
Hi, just out of curiosity, I know opentelemetry plugin now support traces and logging. Is someone working on support OTEL Metric?

**Comment by zzz6519003** [comment]
🥇

**Comment by suzyw-w** [comment]
@zzz6519003  Are you planning working on OTEL metrics support?

**Reply by zzz6519003** [reply]
Just wanna support u

**Comment by suzyw-w** [comment]
@samugi  Hi, I am wondering if I can raise an PR on this? I have some codes for our team internal usage and I'd like to share it with the community.

**Comment by samugi** [comment]
Hi @SuzyWangIBMer , thank you for bringing this discussion to my attention. Contribution is always appreciated! In this particular instance, however, I'd like to know more about your implementation of metrics. We are certainly planning to add support for native OTel metrics, but we are also comparing the efficiency of different solutions to decide the best strategy for supporting this.

If you have any code / links or brief overview to share about this, I would be more than happy to take a look. If you prefer, we can have a chat in Slack at: [Kong Community](https://join.slack.com/t/kongcommunity/shared_invite/zt-1wl3nwzoj-V6PpIz7cbx9u~gqDxmzf_A).

Edit: if you prefer to just go ahead and open the PR, we can continue the discussion there.  Thank you!

**Comment by suzyw-w** [comment]
Hi @samugi  Thanks for prompt response. I don't have access to Kong community slack channel, and the link you shared seems no longer available. We can continue the discussion on slack if you can share the link again. 

My implementation grabs Prometheus metrics value and translate to OLTP format, which is quite straightforward.

I also wanted to open an PR but seems I need write access to create a new branch.

**Comment by samugi** [comment]
my bad @SuzyWangIBMer , please try this: https://join.slack.com/t/kongcommunity/shared_invite/zt-2sntjbx3n-HGtKxcbV2TPxR1HG4mGUkw 

About your implementation, it sounds like something slightly different than what we would be planning, i.e. provide the ability, probably via the PDK, to generate metrics via an integrated metrics system, which could then be used by OTel (and Prometheus, and possibly other plugins) to serialize according to the vendor's spec and report them.

Consider that converting Prometheus metrics to OTLP can be achieved without any code changes using an OpenTelemetry collector, the [Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/README.md) and the [otlp exporter](https://opentelemetry.io/docs/languages/js/exporters/#otlp). Please feel free to reach out on Slack. Thank you!


Edit: about the PR, you can do that by forking the repo and opening the pull request on the upstream from there

**Comment by suzyw-w** [comment]
Hi @samugi, the link works, thanks. [Suzy Wang](https://kongcommunity.slack.com/team/U07RPR8KFPZ)

The team is looking for a way to set up Kong with no other containers to INSTANA (no otlp collector or prometheus receiver). Generating otlp metric via PDK might works for us as well. Can I ask what is the plan on that?

I can go ahead and raise the PR first.
